// React.DOM.SVG/index.js
"use strict";
var React = require("React");
var React_DOM = require("React.DOM");
var React_DOM_Props = require("React.DOM.Props");
var tspan = React_DOM.mkDOM(false)("tspan");
var text = React_DOM.mkDOM(false)("text");
var svg = React_DOM.mkDOM(false)("svg");
var stop = React_DOM.mkDOM(false)("stop");
var rect = React_DOM.mkDOM(false)("rect");
var radialGradient = React_DOM.mkDOM(false)("radialGradient");
var polyline = React_DOM.mkDOM(false)("polyline");
var polygon = React_DOM.mkDOM(false)("polygon");
var pattern = React_DOM.mkDOM(false)("pattern");
var path = React_DOM.mkDOM(false)("path");
var mask = React_DOM.mkDOM(false)("mask");
var linearGradient = React_DOM.mkDOM(false)("linearGradient");
var line = React_DOM.mkDOM(false)("line");
var g = React_DOM.mkDOM(false)("g");
var ellipse = React_DOM.mkDOM(false)("ellipse");
var defs = React_DOM.mkDOM(false)("defs");
var clipPath = React_DOM.mkDOM(false)("clipPath");
var circle = React_DOM.mkDOM(false)("circle");
module.exports = {
    tspan: tspan, 
    text: text, 
    svg: svg, 
    stop: stop, 
    rect: rect, 
    radialGradient: radialGradient, 
    polyline: polyline, 
    polygon: polygon, 
    pattern: pattern, 
    path: path, 
    mask: mask, 
    linearGradient: linearGradient, 
    line: line, 
    g: g, 
    ellipse: ellipse, 
    defs: defs, 
    clipPath: clipPath, 
    circle: circle
};
