// ActionsEffects/index.js
"use strict";
var Control_Monad_Eff_Console = require("Control.Monad.Eff.Console");
var DOM = require("DOM");
var Network_HTTP_Affjax = require("Network.HTTP.Affjax");
var Control_Monad_Eff_Exception = require("Control.Monad.Eff.Exception");
var Prelude = require("Prelude");
var BlogRoutes = require("BlogRoutes");
var Model = require("Model");
var NoFx = (function () {
    function NoFx() {

    };
    NoFx.value = new NoFx();
    return NoFx;
})();
var GoTo = (function () {
    function GoTo(value0) {
        this.value0 = value0;
    };
    GoTo.create = function (value0) {
        return new GoTo(value0);
    };
    return GoTo;
})();
var FetchPosts = (function () {
    function FetchPosts() {

    };
    FetchPosts.value = new FetchPosts();
    return FetchPosts;
})();
var NoOp = (function () {
    function NoOp() {

    };
    NoOp.value = new NoOp();
    return NoOp;
})();
var SetPage = (function () {
    function SetPage(value0) {
        this.value0 = value0;
    };
    SetPage.create = function (value0) {
        return new SetPage(value0);
    };
    return SetPage;
})();
var SetPosts = (function () {
    function SetPosts(value0) {
        this.value0 = value0;
    };
    SetPosts.create = function (value0) {
        return new SetPosts(value0);
    };
    return SetPosts;
})();
var showAction = new Prelude.Show(function (v) {
    if (v instanceof NoOp) {
        return "NoOp";
    };
    if (v instanceof SetPage) {
        return "SetPage";
    };
    if (v instanceof SetPosts) {
        return "SetPosts";
    };
    throw new Error("Failed pattern match at ActionsEffects line 19, column 3 - line 20, column 3: " + [ v.constructor.name ]);
});
module.exports = {
    NoFx: NoFx, 
    GoTo: GoTo, 
    FetchPosts: FetchPosts, 
    NoOp: NoOp, 
    SetPage: SetPage, 
    SetPosts: SetPosts, 
    showAction: showAction
};
