// Components.Pages/index.js
"use strict";
var Prelude = require("Prelude");
var Control_Monad_Eff = require("Control.Monad.Eff");
var Control_Monad_Eff_Console = require("Control.Monad.Eff.Console");
var React = require("React");
var ReactDOM = require("ReactDOM");
var React_DOM = require("React.DOM");
var React_DOM_Props = require("React.DOM.Props");
var Data_RxState = require("Data.RxState");
var DOM_Event_Event = require("DOM.Event.Event");
var DOM_Event_Types = require("DOM.Event.Types");
var Unsafe_Coerce = require("Unsafe.Coerce");
var BlogRoutes = require("BlogRoutes");
var ActionsEffects = require("ActionsEffects");
var Model = require("Model");
var Channels = require("Channels");
var Validators = require("Validators");
var onePost = function (n) {
    return React.createClass(React.spec(n)(function (ctx) {
        return function __do() {
            var v = React.getProps(ctx)();
            return React_DOM.div([  ])([ React_DOM.text("This is the NewPost Page") ]);
        };
    }));
};
var newPost = React.createClass(React.spec(Prelude.unit)(function (ctx) {
    return function __do() {
        var v = React.getProps(ctx)();
        return React_DOM["div'"]([ React_DOM.text("Create a new post") ]);
    };
}));
var indexWillMount = function (v) {
    return Data_RxState.send([ ActionsEffects.FetchPosts.value ])(Channels.effectsChannel);
};
var indexRender = function (ctx) {
    return function __do() {
        var v = React.getProps(ctx)();
        return React_DOM.div([  ])([ React_DOM.div([ React_DOM_Props.className("text-xs-right") ])([ React_DOM.a([ React_DOM_Props.href(Prelude.show(BlogRoutes.routesShow)(BlogRoutes.NewPost.value)), React_DOM_Props.className("btn btn-primary") ])([ React_DOM.text("Add a Post") ]) ]) ]);
    };
};
var indexSpec = {
    render: indexRender, 
    displayName: "index", 
    getInitialState: function (v) {
        return Prelude.pure(Control_Monad_Eff.applicativeEff)(Prelude.unit);
    }, 
    componentWillMount: indexWillMount, 
    componentDidMount: function (v) {
        return Prelude["return"](Control_Monad_Eff.applicativeEff)(Prelude.unit);
    }, 
    componentWillReceiveProps: function (v) {
        return function (v1) {
            return Prelude["return"](Control_Monad_Eff.applicativeEff)(Prelude.unit);
        };
    }, 
    shouldComponentUpdate: function (v) {
        return function (v1) {
            return function (v2) {
                return Prelude["return"](Control_Monad_Eff.applicativeEff)(true);
            };
        };
    }, 
    componentWillUpdate: function (v) {
        return function (v1) {
            return function (v2) {
                return Prelude["return"](Control_Monad_Eff.applicativeEff)(Prelude.unit);
            };
        };
    }, 
    componentDidUpdate: function (v) {
        return function (v1) {
            return function (v2) {
                return Prelude["return"](Control_Monad_Eff.applicativeEff)(Prelude.unit);
            };
        };
    }, 
    componentWillUnmount: function (v) {
        return Prelude["return"](Control_Monad_Eff.applicativeEff)(Prelude.unit);
    }
};
var index = React.createClass(indexSpec);
module.exports = {
    newPost: newPost, 
    onePost: onePost, 
    index: index, 
    indexSpec: indexSpec, 
    indexRender: indexRender, 
    indexWillMount: indexWillMount
};
