// React.DOM.Dynamic/index.js
"use strict";
var React = require("React");
var React_DOM_Props = require("React.DOM.Props");
var React_DOM = require("React.DOM");
var wbr = React_DOM.mkDOM(true)("body");
var wbr$prime = wbr([  ]);
var video = React_DOM.mkDOM(true)("video");
var video$prime = video([  ]);
var $$var = React_DOM.mkDOM(true)("var");
var var$prime = $$var([  ]);
var ul = React_DOM.mkDOM(true)("ul");
var ul$prime = ul([  ]);
var u = React_DOM.mkDOM(true)("u");
var u$prime = u([  ]);
var track = React_DOM.mkDOM(true)("track");
var track$prime = track([  ]);
var tr = React_DOM.mkDOM(true)("tr");
var tr$prime = tr([  ]);
var title = React_DOM.mkDOM(true)("title");
var title$prime = title([  ]);
var time = React_DOM.mkDOM(true)("time");
var time$prime = time([  ]);
var thead = React_DOM.mkDOM(true)("thead");
var thead$prime = thead([  ]);
var th = React_DOM.mkDOM(true)("th");
var th$prime = th([  ]);
var tfoot = React_DOM.mkDOM(true)("tfoot");
var tfoot$prime = tfoot([  ]);
var textarea = React_DOM.mkDOM(true)("textarea");
var textarea$prime = textarea([  ]);
var text = React_DOM.text;
var td = React_DOM.mkDOM(true)("td");
var td$prime = td([  ]);
var tbody = React_DOM.mkDOM(true)("tbody");
var tbody$prime = tbody([  ]);
var table = React_DOM.mkDOM(true)("table");
var table$prime = table([  ]);
var sup = React_DOM.mkDOM(true)("sup");
var sup$prime = sup([  ]);
var summary = React_DOM.mkDOM(true)("summary");
var summary$prime = summary([  ]);
var sub = React_DOM.mkDOM(true)("sub");
var sub$prime = sub([  ]);
var style = React_DOM.mkDOM(true)("style");
var style$prime = style([  ]);
var strong = React_DOM.mkDOM(true)("strong");
var strong$prime = strong([  ]);
var span = React_DOM.mkDOM(true)("span");
var span$prime = span([  ]);
var source = React_DOM.mkDOM(true)("source");
var source$prime = source([  ]);
var small = React_DOM.mkDOM(true)("small");
var small$prime = small([  ]);
var select = React_DOM.mkDOM(true)("select");
var select$prime = select([  ]);
var section = React_DOM.mkDOM(true)("section");
var section$prime = section([  ]);
var script = React_DOM.mkDOM(true)("script");
var script$prime = script([  ]);
var samp = React_DOM.mkDOM(true)("samp");
var samp$prime = samp([  ]);
var s = React_DOM.mkDOM(true)("s");
var s$prime = s([  ]);
var ruby = React_DOM.mkDOM(true)("ruby");
var ruby$prime = ruby([  ]);
var rt = React_DOM.mkDOM(true)("rt");
var rt$prime = rt([  ]);
var rp = React_DOM.mkDOM(true)("rp");
var rp$prime = rp([  ]);
var q = React_DOM.mkDOM(true)("q");
var q$prime = q([  ]);
var progress = React_DOM.mkDOM(true)("progress");
var progress$prime = progress([  ]);
var pre = React_DOM.mkDOM(true)("pre");
var pre$prime = pre([  ]);
var picture = React_DOM.mkDOM(true)("picture");
var picture$prime = picture([  ]);
var param = React_DOM.mkDOM(true)("param");
var param$prime = param([  ]);
var p = React_DOM.mkDOM(true)("p");
var p$prime = p([  ]);
var output = React_DOM.mkDOM(true)("output");
var output$prime = output([  ]);
var option = React_DOM.mkDOM(true)("option");
var option$prime = option([  ]);
var optgroup = React_DOM.mkDOM(true)("optgroup");
var optgroup$prime = optgroup([  ]);
var ol = React_DOM.mkDOM(true)("ol");
var ol$prime = ol([  ]);
var object = React_DOM.mkDOM(true)("object");
var object$prime = object([  ]);
var noscript = React_DOM.mkDOM(true)("noscript");
var noscript$prime = noscript([  ]);
var nav = React_DOM.mkDOM(true)("nav");
var nav$prime = nav([  ]);
var meter = React_DOM.mkDOM(true)("meter");
var meter$prime = meter([  ]);
var meta = React_DOM.mkDOM(true)("meta");
var meta$prime = meta([  ]);
var menuitem = React_DOM.mkDOM(true)("menuitem");
var menuitem$prime = menuitem([  ]);
var menu = React_DOM.mkDOM(true)("menu");
var menu$prime = menu([  ]);
var mark = React_DOM.mkDOM(true)("mark");
var mark$prime = mark([  ]);
var map = React_DOM.mkDOM(true)("map");
var map$prime = map([  ]);
var main = React_DOM.mkDOM(true)("main");
var main$prime = main([  ]);
var link = React_DOM.mkDOM(true)("link");
var li = React_DOM.mkDOM(true)("li");
var li$prime = li([  ]);
var legend = React_DOM.mkDOM(true)("legend");
var legend$prime = legend([  ]);
var label = React_DOM.mkDOM(true)("label");
var label$prime = label([  ]);
var keygen = React_DOM.mkDOM(true)("keygen");
var keygen$prime = keygen([  ]);
var kbd = React_DOM.mkDOM(true)("kbd");
var kbd$prime = kbd([  ]);
var ins = React_DOM.mkDOM(true)("ins");
var ins$prime = ins([  ]);
var input = React_DOM.mkDOM(true)("input");
var input$prime = input([  ]);
var img = React_DOM.mkDOM(true)("img");
var img$prime = img([  ]);
var iframe = React_DOM.mkDOM(true)("iframe");
var iframe$prime = iframe([  ]);
var i = React_DOM.mkDOM(true)("i");
var i$prime = i([  ]);
var html = React_DOM.mkDOM(true)("html");
var html$prime = html([  ]);
var hr = React_DOM.mkDOM(true)("hr");
var hr$prime = hr([  ]);
var header = React_DOM.mkDOM(true)("header");
var header$prime = header([  ]);
var head = React_DOM.mkDOM(true)("head");
var head$prime = head([  ]);
var h6 = React_DOM.mkDOM(true)("h6");
var h6$prime = h6([  ]);
var h5 = React_DOM.mkDOM(true)("h5");
var h5$prime = h5([  ]);
var h4 = React_DOM.mkDOM(true)("h4");
var h4$prime = h4([  ]);
var h3 = React_DOM.mkDOM(true)("h3");
var h3$prime = h3([  ]);
var h2 = React_DOM.mkDOM(true)("h2");
var h2$prime = h2([  ]);
var h1 = React_DOM.mkDOM(true)("h1");
var h1$prime = h1([  ]);
var form = React_DOM.mkDOM(true)("form");
var form$prime = form([  ]);
var footer = React_DOM.mkDOM(true)("footer");
var footer$prime = footer([  ]);
var figure = React_DOM.mkDOM(true)("figure");
var figure$prime = figure([  ]);
var figcaption = React_DOM.mkDOM(true)("figcaption");
var figcaption$prime = figcaption([  ]);
var fieldset = React_DOM.mkDOM(true)("fieldset");
var fieldset$prime = fieldset([  ]);
var embed = React_DOM.mkDOM(true)("embed");
var embed$prime = embed([  ]);
var em = React_DOM.mkDOM(true)("em");
var em$prime = em([  ]);
var dt = React_DOM.mkDOM(true)("dt");
var dt$prime = dt([  ]);
var dl = React_DOM.mkDOM(true)("dl");
var dl$prime = dl([  ]);
var div = React_DOM.mkDOM(true)("div");
var div$prime = div([  ]);
var dialog = React_DOM.mkDOM(true)("dialog");
var dialog$prime = dialog([  ]);
var dfn = React_DOM.mkDOM(true)("dfn");
var dfn$prime = dfn([  ]);
var details = React_DOM.mkDOM(true)("details");
var details$prime = details([  ]);
var del = React_DOM.mkDOM(true)("del");
var del$prime = del([  ]);
var dd = React_DOM.mkDOM(true)("dd");
var dd$prime = dd([  ]);
var datalist = React_DOM.mkDOM(true)("datalist");
var datalist$prime = datalist([  ]);
var colgroup = React_DOM.mkDOM(true)("colgroup");
var colgroup$prime = colgroup([  ]);
var col = React_DOM.mkDOM(true)("col");
var col$prime = col([  ]);
var code = React_DOM.mkDOM(true)("code");
var code$prime = code([  ]);
var cite = React_DOM.mkDOM(true)("cite");
var cite$prime = cite([  ]);
var caption = React_DOM.mkDOM(true)("caption");
var caption$prime = caption([  ]);
var canvas = React_DOM.mkDOM(true)("canvas");
var canvas$prime = canvas([  ]);
var button = React_DOM.mkDOM(true)("button");
var button$prime = button([  ]);
var br = React_DOM.mkDOM(true)("br");
var br$prime = br([  ]);
var body = React_DOM.mkDOM(true)("body");
var body$prime = body([  ]);
var link$prime = body([  ]);
var blockquote = React_DOM.mkDOM(true)("blockquote");
var blockquote$prime = blockquote([  ]);
var big = React_DOM.mkDOM(true)("big");
var big$prime = big([  ]);
var bdo = React_DOM.mkDOM(true)("bdo");
var bdo$prime = bdo([  ]);
var bdi = React_DOM.mkDOM(true)("bdi");
var bdi$prime = bdi([  ]);
var base = React_DOM.mkDOM(true)("base");
var base$prime = base([  ]);
var b = React_DOM.mkDOM(true)("b");
var b$prime = b([  ]);
var audio = React_DOM.mkDOM(true)("audio");
var audio$prime = audio([  ]);
var aside = React_DOM.mkDOM(true)("aside");
var aside$prime = aside([  ]);
var article = React_DOM.mkDOM(true)("article");
var article$prime = article([  ]);
var area = React_DOM.mkDOM(true)("area");
var area$prime = area([  ]);
var address = React_DOM.mkDOM(true)("address");
var address$prime = address([  ]);
var abbr = React_DOM.mkDOM(true)("abbr");
var abbr$prime = abbr([  ]);
var a = React_DOM.mkDOM(true)("a");
var a$prime = a([  ]);
var _data = React_DOM.mkDOM(true)("data");
var _data$prime = _data([  ]);
module.exports = {
    "wbr'": wbr$prime, 
    wbr: wbr, 
    "video'": video$prime, 
    video: video, 
    "var'": var$prime, 
    "var": $$var, 
    "ul'": ul$prime, 
    ul: ul, 
    "u'": u$prime, 
    u: u, 
    "track'": track$prime, 
    track: track, 
    "tr'": tr$prime, 
    tr: tr, 
    "title'": title$prime, 
    title: title, 
    "time'": time$prime, 
    time: time, 
    "thead'": thead$prime, 
    thead: thead, 
    "th'": th$prime, 
    th: th, 
    "tfoot'": tfoot$prime, 
    tfoot: tfoot, 
    "textarea'": textarea$prime, 
    textarea: textarea, 
    "td'": td$prime, 
    td: td, 
    "tbody'": tbody$prime, 
    tbody: tbody, 
    "table'": table$prime, 
    table: table, 
    "sup'": sup$prime, 
    sup: sup, 
    "summary'": summary$prime, 
    summary: summary, 
    "sub'": sub$prime, 
    sub: sub, 
    "style'": style$prime, 
    style: style, 
    "strong'": strong$prime, 
    strong: strong, 
    "span'": span$prime, 
    span: span, 
    "source'": source$prime, 
    source: source, 
    "small'": small$prime, 
    small: small, 
    "select'": select$prime, 
    select: select, 
    "section'": section$prime, 
    section: section, 
    "script'": script$prime, 
    script: script, 
    "samp'": samp$prime, 
    samp: samp, 
    "s'": s$prime, 
    s: s, 
    "ruby'": ruby$prime, 
    ruby: ruby, 
    "rt'": rt$prime, 
    rt: rt, 
    "rp'": rp$prime, 
    rp: rp, 
    "q'": q$prime, 
    q: q, 
    "progress'": progress$prime, 
    progress: progress, 
    "pre'": pre$prime, 
    pre: pre, 
    "picture'": picture$prime, 
    picture: picture, 
    "param'": param$prime, 
    param: param, 
    "p'": p$prime, 
    p: p, 
    "output'": output$prime, 
    output: output, 
    "option'": option$prime, 
    option: option, 
    "optgroup'": optgroup$prime, 
    optgroup: optgroup, 
    "ol'": ol$prime, 
    ol: ol, 
    "object'": object$prime, 
    object: object, 
    "noscript'": noscript$prime, 
    noscript: noscript, 
    "nav'": nav$prime, 
    nav: nav, 
    "meter'": meter$prime, 
    meter: meter, 
    "meta'": meta$prime, 
    meta: meta, 
    "menuitem'": menuitem$prime, 
    menuitem: menuitem, 
    "menu'": menu$prime, 
    menu: menu, 
    "mark'": mark$prime, 
    mark: mark, 
    "map'": map$prime, 
    map: map, 
    "main'": main$prime, 
    main: main, 
    "link'": link$prime, 
    link: link, 
    "li'": li$prime, 
    li: li, 
    "legend'": legend$prime, 
    legend: legend, 
    "label'": label$prime, 
    label: label, 
    "keygen'": keygen$prime, 
    keygen: keygen, 
    "kbd'": kbd$prime, 
    kbd: kbd, 
    "ins'": ins$prime, 
    ins: ins, 
    "input'": input$prime, 
    input: input, 
    "img'": img$prime, 
    img: img, 
    "iframe'": iframe$prime, 
    iframe: iframe, 
    "i'": i$prime, 
    i: i, 
    "html'": html$prime, 
    html: html, 
    "hr'": hr$prime, 
    hr: hr, 
    "header'": header$prime, 
    header: header, 
    "head'": head$prime, 
    head: head, 
    "h6'": h6$prime, 
    h6: h6, 
    "h5'": h5$prime, 
    h5: h5, 
    "h4'": h4$prime, 
    h4: h4, 
    "h3'": h3$prime, 
    h3: h3, 
    "h2'": h2$prime, 
    h2: h2, 
    "h1'": h1$prime, 
    h1: h1, 
    "form'": form$prime, 
    form: form, 
    "footer'": footer$prime, 
    footer: footer, 
    "figure'": figure$prime, 
    figure: figure, 
    "figcaption'": figcaption$prime, 
    figcaption: figcaption, 
    "fieldset'": fieldset$prime, 
    fieldset: fieldset, 
    "embed'": embed$prime, 
    embed: embed, 
    "em'": em$prime, 
    em: em, 
    "dt'": dt$prime, 
    dt: dt, 
    "dl'": dl$prime, 
    dl: dl, 
    "div'": div$prime, 
    div: div, 
    "dialog'": dialog$prime, 
    dialog: dialog, 
    "dfn'": dfn$prime, 
    dfn: dfn, 
    "details'": details$prime, 
    details: details, 
    "del'": del$prime, 
    del: del, 
    "dd'": dd$prime, 
    dd: dd, 
    "datalist'": datalist$prime, 
    datalist: datalist, 
    "_data'": _data$prime, 
    _data: _data, 
    "colgroup'": colgroup$prime, 
    colgroup: colgroup, 
    "col'": col$prime, 
    col: col, 
    "code'": code$prime, 
    code: code, 
    "cite'": cite$prime, 
    cite: cite, 
    "caption'": caption$prime, 
    caption: caption, 
    "canvas'": canvas$prime, 
    canvas: canvas, 
    "button'": button$prime, 
    button: button, 
    "br'": br$prime, 
    br: br, 
    "body'": body$prime, 
    body: body, 
    "blockquote'": blockquote$prime, 
    blockquote: blockquote, 
    "big'": big$prime, 
    big: big, 
    "bdo'": bdo$prime, 
    bdo: bdo, 
    "bdi'": bdi$prime, 
    bdi: bdi, 
    "base'": base$prime, 
    base: base, 
    "b'": b$prime, 
    b: b, 
    "audio'": audio$prime, 
    audio: audio, 
    "aside'": aside$prime, 
    aside: aside, 
    "article'": article$prime, 
    article: article, 
    "area'": area$prime, 
    area: area, 
    "address'": address$prime, 
    address: address, 
    "abbr'": abbr$prime, 
    abbr: abbr, 
    "a'": a$prime, 
    a: a, 
    text: text
};
