// React.DOM.SVG.Dynamic/index.js
"use strict";
var React = require("React");
var React_DOM = require("React.DOM");
var React_DOM_Props = require("React.DOM.Props");
var tspan = React_DOM.mkDOM(true)("tspan");
var text = React_DOM.mkDOM(true)("text");
var svg = React_DOM.mkDOM(true)("svg");
var stop = React_DOM.mkDOM(true)("stop");
var rect = React_DOM.mkDOM(true)("rect");
var radialGradient = React_DOM.mkDOM(true)("radialGradient");
var polyline = React_DOM.mkDOM(true)("polyline");
var polygon = React_DOM.mkDOM(true)("polygon");
var pattern = React_DOM.mkDOM(true)("pattern");
var path = React_DOM.mkDOM(true)("path");
var mask = React_DOM.mkDOM(true)("mask");
var linearGradient = React_DOM.mkDOM(true)("linearGradient");
var line = React_DOM.mkDOM(true)("line");
var g = React_DOM.mkDOM(true)("g");
var ellipse = React_DOM.mkDOM(true)("ellipse");
var defs = React_DOM.mkDOM(true)("defs");
var clipPath = React_DOM.mkDOM(true)("clipPath");
var circle = React_DOM.mkDOM(true)("circle");
module.exports = {
    tspan: tspan, 
    text: text, 
    svg: svg, 
    stop: stop, 
    rect: rect, 
    radialGradient: radialGradient, 
    polyline: polyline, 
    polygon: polygon, 
    pattern: pattern, 
    path: path, 
    mask: mask, 
    linearGradient: linearGradient, 
    line: line, 
    g: g, 
    ellipse: ellipse, 
    defs: defs, 
    clipPath: clipPath, 
    circle: circle
};
