// Data.Argonaut.JCursor/index.js
"use strict";
var $foreign = require("./foreign");
var Prelude = require("Prelude");
var Data_Argonaut_Core = require("Data.Argonaut.Core");
var Data_Argonaut_Decode = require("Data.Argonaut.Decode");
var Data_Argonaut_Encode = require("Data.Argonaut.Encode");
var Data_Either = require("Data.Either");
var Data_Foldable = require("Data.Foldable");
var Data_List = require("Data.List");
var Data_Maybe = require("Data.Maybe");
var Data_Monoid = require("Data.Monoid");
var Data_Tuple = require("Data.Tuple");
var Data_Array = require("Data.Array");
var Data_Int = require("Data.Int");
var Data_Maybe_Unsafe = require("Data.Maybe.Unsafe");
var Data_StrMap = require("Data.StrMap");
var JsonPrim = function (x) {
    return x;
};
var JCursorTop = (function () {
    function JCursorTop() {

    };
    JCursorTop.value = new JCursorTop();
    return JCursorTop;
})();
var JField = (function () {
    function JField(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    JField.create = function (value0) {
        return function (value1) {
            return new JField(value0, value1);
        };
    };
    return JField;
})();
var JIndex = (function () {
    function JIndex(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    JIndex.create = function (value0) {
        return function (value1) {
            return new JIndex(value0, value1);
        };
    };
    return JIndex;
})();
var showJCursor = new Prelude.Show(function (v) {
    if (v instanceof JCursorTop) {
        return "";
    };
    if (v instanceof JField) {
        return "." + (v.value0 + Prelude.show(showJCursor)(v.value1));
    };
    if (v instanceof JIndex) {
        return "[" + (Prelude.show(Prelude.showInt)(v.value0) + ("]" + Prelude.show(showJCursor)(v.value1)));
    };
    throw new Error("Failed pattern match at Data.Argonaut.JCursor line 158, column 3 - line 159, column 3: " + [ v.constructor.name ]);
});
var semigroupJCursor = new Prelude.Semigroup(function (v) {
    return function (v1) {
        if (v1 instanceof JCursorTop) {
            return v;
        };
        if (v instanceof JCursorTop) {
            return v1;
        };
        if (v instanceof JField) {
            return new JField(v.value0, Prelude["<>"](semigroupJCursor)(v.value1)(v1));
        };
        if (v instanceof JIndex) {
            return new JIndex(v.value0, Prelude["<>"](semigroupJCursor)(v.value1)(v1));
        };
        throw new Error("Failed pattern match at Data.Argonaut.JCursor line 185, column 3 - line 186, column 3: " + [ v.constructor.name, v1.constructor.name ]);
    };
});
var runJsonPrim = function (v) {
    return v;
};
var showJsonPrim = new Prelude.Show(function (p) {
    return runJsonPrim(p)(Prelude.show(Data_Argonaut_Core.showJNull))(Prelude.show(Prelude.showBoolean))(Prelude.show(Prelude.showNumber))(Prelude.show(Prelude.showString));
});
var primToJson = function (p) {
    return runJsonPrim(p)(Data_Argonaut_Core.fromNull)(Data_Argonaut_Core.fromBoolean)(Data_Argonaut_Core.fromNumber)(Data_Argonaut_Core.fromString);
};
var primStr = function (v) {
    return function (v1) {
        return function (v2) {
            return function (v3) {
                return function (f) {
                    return f(v);
                };
            };
        };
    };
};
var primNum = function (v) {
    return function (v1) {
        return function (v2) {
            return function (f) {
                return function (v3) {
                    return f(v);
                };
            };
        };
    };
};
var primNull = function (f) {
    return function (v) {
        return function (v1) {
            return function (v2) {
                return f($foreign.exactNull);
            };
        };
    };
};
var primBool = function (v) {
    return function (v1) {
        return function (f) {
            return function (v2) {
                return function (v3) {
                    return f(v);
                };
            };
        };
    };
};
var toPrims = (function () {
    var objFn = function (obj) {
        var f = function (v) {
            return Prelude["<$>"](Data_List.functorList)(function (t) {
                return new Data_Tuple.Tuple(new JField(v.value0, Data_Tuple.fst(t)), Data_Tuple.snd(t));
            })(toPrims(v.value1));
        };
        return Prelude[">>="](Data_List.bindList)(Data_StrMap.toList(obj))(f);
    };
    var mkTop = function (p) {
        return Data_List.singleton(new Data_Tuple.Tuple(JCursorTop.value, p));
    };
    var nullFn = function (v) {
        return mkTop(primNull);
    };
    var numFn = function (n) {
        return mkTop(primNum(n));
    };
    var strFn = function (s) {
        return mkTop(primStr(s));
    };
    var boolFn = function (b) {
        return mkTop(primBool(b));
    };
    var arrFn$prime = function (v) {
        return Data_List.toList(Data_List.foldableList)(Prelude["<$>"](Data_List.functorList)(function (t) {
            return new Data_Tuple.Tuple(new JIndex(v.value0, Data_Tuple.fst(t)), Data_Tuple.snd(t));
        })(toPrims(v.value1)));
    };
    var arrFn = function (arr) {
        var zipped = Data_List.zipWith(Data_Tuple.Tuple.create)(Data_List.range(0)(Data_Array.length(arr) - 1))(Data_List.toList(Data_Foldable.foldableArray)(arr));
        return Prelude[">>="](Data_List.bindList)(zipped)(arrFn$prime);
    };
    return Data_Argonaut_Core.foldJson(nullFn)(boolFn)(numFn)(strFn)(arrFn)(objFn);
})();
var monoidJCursor = new Data_Monoid.Monoid(function () {
    return semigroupJCursor;
}, JCursorTop.value);
var inferEmpty = function (v) {
    if (v instanceof JCursorTop) {
        return Data_Argonaut_Core.jsonNull;
    };
    if (v instanceof JField) {
        return Data_Argonaut_Core.jsonEmptyObject;
    };
    if (v instanceof JIndex) {
        return Data_Argonaut_Core.jsonEmptyArray;
    };
    throw new Error("Failed pattern match at Data.Argonaut.JCursor line 78, column 1 - line 79, column 1: " + [ v.constructor.name ]);
};
var fail = function (dictShow) {
    return function (x) {
        return Data_Either.Left.create("Expected String or Number but found: " + Prelude.show(dictShow)(x));
    };
};
var eqJCursor = new Prelude.Eq(function (v) {
    return function (v1) {
        if (v instanceof JCursorTop && v1 instanceof JCursorTop) {
            return true;
        };
        if (v instanceof JField && v1 instanceof JField) {
            return v.value0 === v1.value0 && Prelude["=="](eqJCursor)(v.value1)(v1.value1);
        };
        if (v instanceof JIndex && v1 instanceof JIndex) {
            return v.value0 === v1.value0 && Prelude["=="](eqJCursor)(v.value1)(v1.value1);
        };
        return false;
    };
});
var ordJCursor = new Prelude.Ord(function () {
    return eqJCursor;
}, function (v) {
    return function (v1) {
        if (v instanceof JCursorTop && v1 instanceof JCursorTop) {
            return Prelude.EQ.value;
        };
        if (v instanceof JCursorTop) {
            return Prelude.LT.value;
        };
        if (v1 instanceof JCursorTop) {
            return Prelude.GT.value;
        };
        if (v instanceof JField && v1 instanceof JIndex) {
            return Prelude.LT.value;
        };
        if (v instanceof JIndex && v1 instanceof JField) {
            return Prelude.GT.value;
        };
        if (v instanceof JField && v1 instanceof JField) {
            var $91 = Prelude.compare(Prelude.ordString)(v.value0)(v1.value0);
            if ($91 instanceof Prelude.EQ) {
                return Prelude.compare(ordJCursor)(v.value1)(v1.value1);
            };
            return $91;
        };
        if (v instanceof JIndex && v1 instanceof JIndex) {
            var $96 = Prelude.compare(Prelude.ordInt)(v.value0)(v1.value0);
            if ($96 instanceof Prelude.EQ) {
                return Prelude.compare(ordJCursor)(v.value1)(v1.value1);
            };
            return $96;
        };
        throw new Error("Failed pattern match at Data.Argonaut.JCursor line 172, column 3 - line 173, column 3: " + [ v.constructor.name, v1.constructor.name ]);
    };
});
var encodeJsonJCursor = new Data_Argonaut_Encode.EncodeJson((function () {
    var loop = function (v) {
        if (v instanceof JCursorTop) {
            return [  ];
        };
        if (v instanceof JField) {
            return Prelude["<>"](Prelude.semigroupArray)([ Data_Argonaut_Encode.encodeJson(Data_Argonaut_Encode.encodeJsonJString)(v.value0) ])(loop(v.value1));
        };
        if (v instanceof JIndex) {
            return Prelude["<>"](Prelude.semigroupArray)([ Data_Argonaut_Encode.encodeJson(Data_Argonaut_Encode.encodeJsonInt)(v.value0) ])(loop(v.value1));
        };
        throw new Error("Failed pattern match at Data.Argonaut.JCursor line 194, column 3 - line 199, column 1: " + [ v.constructor.name ]);
    };
    return function ($144) {
        return Data_Argonaut_Encode.encodeJson(Data_Argonaut_Encode.encodeJsonArray(Data_Argonaut_Encode.encodeJsonJson))(loop($144));
    };
})());
var downIndex = function (i) {
    var downIndex$prime = function (v) {
        if (v instanceof JCursorTop) {
            return new JIndex(i, JCursorTop.value);
        };
        if (v instanceof JField) {
            return new JField(v.value0, downIndex$prime(v.value1));
        };
        if (v instanceof JIndex) {
            return new JIndex(v.value0, downIndex$prime(v.value1));
        };
        throw new Error("Failed pattern match at Data.Argonaut.JCursor line 65, column 1 - line 70, column 1: " + [ v.constructor.name ]);
    };
    return downIndex$prime;
};
var downField = function (i) {
    var downField$prime = function (v) {
        if (v instanceof JCursorTop) {
            return new JField(i, JCursorTop.value);
        };
        if (v instanceof JField) {
            return new JField(v.value0, downField$prime(v.value1));
        };
        if (v instanceof JIndex) {
            return new JIndex(v.value0, downField$prime(v.value1));
        };
        throw new Error("Failed pattern match at Data.Argonaut.JCursor line 59, column 1 - line 64, column 1: " + [ v.constructor.name ]);
    };
    return downField$prime;
};
var insideOut = function (v) {
    if (v instanceof JCursorTop) {
        return JCursorTop.value;
    };
    if (v instanceof JField) {
        return downField(v.value0)(insideOut(v.value1));
    };
    if (v instanceof JIndex) {
        return downIndex(v.value0)(insideOut(v.value1));
    };
    throw new Error("Failed pattern match at Data.Argonaut.JCursor line 54, column 1 - line 55, column 1: " + [ v.constructor.name ]);
};
var decodeJsonJCursor = new Data_Argonaut_Decode.DecodeJson(function (j) {
    var goNum = function (c) {
        return function (num) {
            return Data_Maybe.maybe(new Data_Either.Left("Not an Int"))(function ($145) {
                return Data_Either.Right.create(Prelude.flip(JIndex.create)(c)($145));
            })(Data_Int.fromNumber(num));
        };
    };
    var loop = function (arr) {
        return Data_Maybe.maybe(new Data_Either.Right(JCursorTop.value))(goLoop)(Prelude.bind(Data_Maybe.bindMaybe)(Data_Array.head(arr))(function (v) {
            return Prelude.bind(Data_Maybe.bindMaybe)(Data_Array.tail(arr))(function (v1) {
                return Prelude.pure(Data_Maybe.applicativeMaybe)(new Data_Tuple.Tuple(v, v1));
            });
        }));
    };
    var goLoop = function (v) {
        return Prelude.bind(Data_Either.bindEither)(loop(v.value1))(function (v1) {
            return Data_Argonaut_Core.foldJson(fail(Data_Argonaut_Core.showJNull))(fail(Prelude.showBoolean))(goNum(v1))(function ($146) {
                return Data_Either.Right.create(Prelude.flip(JField.create)(v1)($146));
            })(fail(Prelude.showArray(Data_Argonaut_Core.showJson)))(fail(Data_StrMap.showStrMap(Data_Argonaut_Core.showJson)))(v.value0);
        });
    };
    return Prelude[">>="](Data_Either.bindEither)(Data_Argonaut_Decode.decodeJson(Data_Argonaut_Decode.decodeArray(Data_Argonaut_Decode.decodeJsonJson))(j))(loop);
});
var cursorSet = function (v) {
    return function (v1) {
        if (v instanceof JCursorTop) {
            return function ($147) {
                return Prelude.pure(Data_Maybe.applicativeMaybe)(Prelude["const"](v1)($147));
            };
        };
        if (v instanceof JField) {
            var mergeObjs = function (m) {
                return Prelude["<$>"](Data_Maybe.functorMaybe)(function ($148) {
                    return Data_Argonaut_Core.fromObject(Prelude.flip(Data_StrMap.insert(v.value0))(m)($148));
                })(cursorSet(v.value1)(v1)(Data_Maybe.fromMaybe(inferEmpty(v.value1))(Data_StrMap.lookup(v.value0)(m))));
            };
            var defaultObj = Prelude["<$>"](Data_Maybe.functorMaybe)(function ($149) {
                return Data_Argonaut_Core.fromObject(Data_StrMap.singleton(v.value0)($149));
            })(cursorSet(v.value1)(v1)(inferEmpty(v.value1)));
            return Data_Argonaut_Core.foldJsonObject(defaultObj)(mergeObjs);
        };
        if (v instanceof JIndex) {
            var setArr = function (__copy_xs) {
                return function (__copy_i1) {
                    return function (__copy_v3) {
                        var xs = __copy_xs;
                        var i1 = __copy_i1;
                        var v3 = __copy_v3;
                        tco: while (true) {
                            var len = Data_Array.length(xs);
                            var $131 = i1 < 0;
                            if ($131) {
                                return Data_Maybe.Nothing.value;
                            };
                            if (!$131) {
                                var $132 = i1 >= len;
                                if ($132) {
                                    var __tco_xs = Prelude["<>"](Prelude.semigroupArray)(xs)(Data_Array.replicate((i1 - len) + 1 | 0)(Data_Argonaut_Core.jsonNull));
                                    var __tco_i1 = i1;
                                    var __tco_v3 = v3;
                                    xs = __tco_xs;
                                    i1 = __tco_i1;
                                    v3 = __tco_v3;
                                    continue tco;
                                };
                                if (!$132) {
                                    return Data_Maybe.Just.create(Data_Argonaut_Core.fromArray(Data_Maybe_Unsafe.fromJust(Data_Array.updateAt(i1)(v3)(xs))));
                                };
                                throw new Error("Failed pattern match at Data.Argonaut.JCursor line 108, column 13 - line 113, column 1: " + [ $132.constructor.name ]);
                            };
                            throw new Error("Failed pattern match at Data.Argonaut.JCursor line 106, column 8 - line 113, column 1: " + [ $131.constructor.name ]);
                        };
                    };
                };
            };
            var mergeArrs = function (a) {
                return Prelude[">>="](Data_Maybe.bindMaybe)(cursorSet(v.value1)(v1)(Data_Maybe.fromMaybe(inferEmpty(v.value1))(Data_Array["!!"](a)(v.value0))))(setArr(a)(v.value0));
            };
            var defaultArr = Prelude["<$>"](Data_Maybe.functorMaybe)(function ($150) {
                return Data_Argonaut_Core.fromArray(Data_Maybe_Unsafe.fromJust(Prelude.flip(Data_Array.updateAt(v.value0))(Data_Array.replicate(v.value0 + 1 | 0)(Data_Argonaut_Core.jsonNull))($150)));
            })(cursorSet(v.value1)(v1)(inferEmpty(v.value1)));
            return Data_Argonaut_Core.foldJsonArray(defaultArr)(mergeArrs);
        };
        throw new Error("Failed pattern match at Data.Argonaut.JCursor line 83, column 1 - line 84, column 1: " + [ v.constructor.name, v1.constructor.name ]);
    };
};
var fromPrims = function (lst) {
    var f = function (j) {
        return function (v) {
            return Prelude[">>="](Data_Maybe.bindMaybe)(j)(cursorSet(v.value0)(primToJson(v.value1)));
        };
    };
    return Data_Foldable.foldl(Data_List.foldableList)(f)(Prelude["<$>"](Data_Maybe.functorMaybe)(function ($151) {
        return inferEmpty(Data_Tuple.fst($151));
    })(Data_List.head(lst)))(lst);
};
var cursorGet = function (v) {
    if (v instanceof JCursorTop) {
        return Data_Maybe.Just.create;
    };
    if (v instanceof JField) {
        var g = function (m) {
            return Prelude[">>="](Data_Maybe.bindMaybe)(Data_StrMap.lookup(v.value0)(m))(cursorGet(v.value1));
        };
        return Data_Argonaut_Core.foldJsonObject(Data_Maybe.Nothing.value)(g);
    };
    if (v instanceof JIndex) {
        var g = function (a) {
            return Prelude[">>="](Data_Maybe.bindMaybe)(Data_Array["!!"](a)(v.value0))(cursorGet(v.value1));
        };
        return Data_Argonaut_Core.foldJsonArray(Data_Maybe.Nothing.value)(g);
    };
    throw new Error("Failed pattern match at Data.Argonaut.JCursor line 71, column 1 - line 72, column 1: " + [ v.constructor.name ]);
};
module.exports = {
    JsonPrim: JsonPrim, 
    JCursorTop: JCursorTop, 
    JField: JField, 
    JIndex: JIndex, 
    fail: fail, 
    fromPrims: fromPrims, 
    toPrims: toPrims, 
    cursorSet: cursorSet, 
    inferEmpty: inferEmpty, 
    cursorGet: cursorGet, 
    downIndex: downIndex, 
    downField: downField, 
    insideOut: insideOut, 
    primToJson: primToJson, 
    primStr: primStr, 
    primNum: primNum, 
    primBool: primBool, 
    primNull: primNull, 
    runJsonPrim: runJsonPrim, 
    showJCursor: showJCursor, 
    showJsonPrim: showJsonPrim, 
    eqJCursor: eqJCursor, 
    ordJCursor: ordJCursor, 
    semigroupJCursor: semigroupJCursor, 
    monoidJCursor: monoidJCursor, 
    encodeJsonJCursor: encodeJsonJCursor, 
    decodeJsonJCursor: decodeJsonJCursor, 
    exactNull: $foreign.exactNull
};
