// Data.Profunctor/index.js
"use strict";
var Prelude = require("Prelude");

/**
 *  | A `Profunctor` is a `Functor` from the pair category `(Type^op, Type)`
 *  | to `Type`.
 *  |
 *  | In other words, a `Profunctor` is a type constructor of two type
 *  | arguments, which is contravariant in its first argument and covariant
 *  | in its second argument.
 *  |
 *  | The `dimap` function can be used to map functions over both arguments
 *  | simultaneously.
 *  |
 *  | A straightforward example of a profunctor is the function arrow `(->)`.
 *  |
 *  | Laws:
 *  |
 *  | - Identity: `dimap id id = id`
 *  | - Composition: `dimap f1 g1 <<< dimap f2 g2 = dimap (f1 >>> f2) (g1 <<< g2)`
 */
var Profunctor = function (dimap) {
    this.dimap = dimap;
};
var profunctorFn = new Profunctor(function (a2b) {
    return function (c2d) {
        return function (b2c) {
            return function ($4) {
                return c2d(b2c(a2b($4)));
            };
        };
    };
});

/**
 *  | A `Profunctor` is a `Functor` from the pair category `(Type^op, Type)`
 *  | to `Type`.
 *  |
 *  | In other words, a `Profunctor` is a type constructor of two type
 *  | arguments, which is contravariant in its first argument and covariant
 *  | in its second argument.
 *  |
 *  | The `dimap` function can be used to map functions over both arguments
 *  | simultaneously.
 *  |
 *  | A straightforward example of a profunctor is the function arrow `(->)`.
 *  |
 *  | Laws:
 *  |
 *  | - Identity: `dimap id id = id`
 *  | - Composition: `dimap f1 g1 <<< dimap f2 g2 = dimap (f1 >>> f2) (g1 <<< g2)`
 */
var dimap = function (dict) {
    return dict.dimap;
};

/**
 *  | Map a function over the (contravariant) first type argument only.
 */
var lmap = function (dictProfunctor) {
    return function (a2b) {
        return dimap(dictProfunctor)(a2b)(Prelude.id(Prelude.categoryFn));
    };
};

/**
 *  | Map a function over the (covariant) second type argument only.
 */
var rmap = function (dictProfunctor) {
    return function (b2c) {
        return dimap(dictProfunctor)(Prelude.id(Prelude.categoryFn))(b2c);
    };
};

/**
 *  | Lift a pure function into any `Profunctor` which is also a `Category`.
 */
var arr = function (dictCategory) {
    return function (dictProfunctor) {
        return function (f) {
            return rmap(dictProfunctor)(f)(Prelude.id(dictCategory));
        };
    };
};
module.exports = {
    Profunctor: Profunctor, 
    arr: arr, 
    rmap: rmap, 
    lmap: lmap, 
    dimap: dimap, 
    profunctorFn: profunctorFn
};
