// Data.Validation.Semiring/index.js

/**
 *  | This module defines a variant of applicative validation with 
 *  | an `Alternative` instance, for validators which support errors
 *  | with multiple alternatives.
 *  |
 *  | The API is equivalent to `Data.Validation`,
 *  | but uses `Semiring` instead of `Semigroup`.
 */
"use strict";
var Prelude = require("Prelude");
var Control_Alt = require("Control.Alt");
var Control_Plus = require("Control.Plus");
var Control_Alternative = require("Control.Alternative");

/**
 *  | The `V` functor, used for alternative validation
 *  |
 *  | The `Alternative` instance collects multiple failures in
 *  | an arbitrary `Semiring`.
 *  | 
 *  | For example:
 *  | 
 *  | ```purescript
 *  | import Data.Semiring.Free
 *  | 
 *  | validate r :: Person -> V (Free Error) Person 
 *  | validate person = { first: _, last: _, contact: _}
 *  |   <$> validateName person.first
 *  |   <*> validateName person.last
 *  |   <*> (validateEmail person.contact <|> validatePhone person.contact)
 *  | ```
 */
var Valid = (function () {
    function Valid(value0) {
        this.value0 = value0;
    };
    Valid.create = function (value0) {
        return new Valid(value0);
    };
    return Valid;
})();

/**
 *  | The `V` functor, used for alternative validation
 *  |
 *  | The `Alternative` instance collects multiple failures in
 *  | an arbitrary `Semiring`.
 *  | 
 *  | For example:
 *  | 
 *  | ```purescript
 *  | import Data.Semiring.Free
 *  | 
 *  | validate r :: Person -> V (Free Error) Person 
 *  | validate person = { first: _, last: _, contact: _}
 *  |   <$> validateName person.first
 *  |   <*> validateName person.last
 *  |   <*> (validateEmail person.contact <|> validatePhone person.contact)
 *  | ```
 */
var Invalid = (function () {
    function Invalid(value0) {
        this.value0 = value0;
    };
    Invalid.create = function (value0) {
        return new Invalid(value0);
    };
    return Invalid;
})();
var showV = function (dictShow) {
    return function (dictShow1) {
        return new Prelude.Show(function (v) {
            if (v instanceof Invalid) {
                return "Invalid (" + (Prelude.show(dictShow)(v.value0) + ")");
            };
            if (v instanceof Valid) {
                return "Valid (" + (Prelude.show(dictShow1)(v.value0) + ")");
            };
            throw new Error("Failed pattern match at Data.Validation.Semiring line 54, column 3 - line 55, column 3: " + [ v.constructor.name ]);
        });
    };
};

/**
 *  | Unpack the `V` type constructor, providing functions to handle the error
 *  | and success cases.
 */
var runV = function (v) {
    return function (v1) {
        return function (v2) {
            if (v2 instanceof Invalid) {
                return v(v2.value0);
            };
            if (v2 instanceof Valid) {
                return v1(v2.value0);
            };
            throw new Error("Failed pattern match at Data.Validation.Semiring line 45, column 1 - line 46, column 1: " + [ v.constructor.name, v1.constructor.name, v2.constructor.name ]);
        };
    };
};

/**
 *  | Test whether validation was successful or not
 */
var isValid = function (v) {
    if (v instanceof Valid) {
        return true;
    };
    return false;
};

/**
 *  | Fail with a validation error
 */
var invalid = Invalid.create;
var functorV = new Prelude.Functor(function (v) {
    return function (v1) {
        if (v1 instanceof Invalid) {
            return new Invalid(v1.value0);
        };
        if (v1 instanceof Valid) {
            return new Valid(v(v1.value0));
        };
        throw new Error("Failed pattern match at Data.Validation.Semiring line 58, column 3 - line 59, column 3: " + [ v.constructor.name, v1.constructor.name ]);
    };
});
var applyV = function (dictSemiring) {
    return new Prelude.Apply(function () {
        return functorV;
    }, function (v) {
        return function (v1) {
            if (v instanceof Invalid && v1 instanceof Invalid) {
                return new Invalid(Prelude["*"](dictSemiring)(v.value0)(v1.value0));
            };
            if (v instanceof Invalid) {
                return new Invalid(v.value0);
            };
            if (v1 instanceof Invalid) {
                return new Invalid(v1.value0);
            };
            if (v instanceof Valid && v1 instanceof Valid) {
                return new Valid(v.value0(v1.value0));
            };
            throw new Error("Failed pattern match at Data.Validation.Semiring line 62, column 3 - line 63, column 3: " + [ v.constructor.name, v1.constructor.name ]);
        };
    });
};
var applicativeV = function (dictSemiring) {
    return new Prelude.Applicative(function () {
        return applyV(dictSemiring);
    }, Valid.create);
};
var altV = function (dictSemiring) {
    return new Control_Alt.Alt(function () {
        return functorV;
    }, function (v) {
        return function (v1) {
            if (v instanceof Invalid && v1 instanceof Invalid) {
                return new Invalid(Prelude["+"](dictSemiring)(v.value0)(v1.value0));
            };
            if (v instanceof Invalid) {
                return v1;
            };
            if (v instanceof Valid) {
                return new Valid(v.value0);
            };
            throw new Error("Failed pattern match at Data.Validation.Semiring line 71, column 3 - line 72, column 3: " + [ v.constructor.name, v1.constructor.name ]);
        };
    });
};
var plusV = function (dictSemiring) {
    return new Control_Plus.Plus(function () {
        return altV(dictSemiring);
    }, new Invalid(Prelude.zero(dictSemiring)));
};
var alernativeV = function (dictSemiring) {
    return new Control_Alternative.Alternative(function () {
        return plusV(dictSemiring);
    }, function () {
        return applicativeV(dictSemiring);
    });
};
module.exports = {
    isValid: isValid, 
    runV: runV, 
    invalid: invalid, 
    showV: showV, 
    functorV: functorV, 
    applyV: applyV, 
    applicativeV: applicativeV, 
    altV: altV, 
    plusV: plusV, 
    alernativeV: alernativeV
};
