// Model/index.js
"use strict";
var Prelude = require("Prelude");
var Data_Argonaut_Core = require("Data.Argonaut.Core");
var Data_Argonaut_Combinators = require("Data.Argonaut.Combinators");
var Data_Argonaut_Encode = require("Data.Argonaut.Encode");
var Data_Argonaut_Decode = require("Data.Argonaut.Decode");
var Data_Maybe = require("Data.Maybe");
var Data_Either = require("Data.Either");
var Data_Traversable = require("Data.Traversable");
var BlogRoutes = require("BlogRoutes");
var BlogPost = function (x) {
    return x;
};
var initialModel = {
    currentPost: Data_Maybe.Nothing.value, 
    allPosts: [  ], 
    currPage: BlogRoutes.Index.value
};
var encodeJson = new Data_Argonaut_Encode.EncodeJson(function (v) {
    return Data_Argonaut_Combinators["~>"](Data_Argonaut_Encode.encodeJsonJson)(Data_Argonaut_Combinators[":="](Data_Argonaut_Encode.encodeJsonInt)("id")(v.id))(Data_Argonaut_Combinators["~>"](Data_Argonaut_Encode.encodeJsonJson)(Data_Argonaut_Combinators[":="](Data_Argonaut_Encode.encodeJsonJString)("title")(v.title))(Data_Argonaut_Combinators["~>"](Data_Argonaut_Encode.encodeJsonJson)(Data_Argonaut_Combinators[":="](Data_Argonaut_Encode.encodeJsonJString)("categories")(v.categories))(Data_Argonaut_Combinators["~>"](Data_Argonaut_Encode.encodeJsonJson)(Data_Argonaut_Combinators[":="](Data_Argonaut_Encode.encodeJsonJString)("content")(v.content))(Data_Argonaut_Core.jsonEmptyObject))));
});
var encodeBlogPostArray = function (bpa) {
    return Data_Argonaut_Core.fromArray(Prelude["<$>"](Prelude.functorArray)(Data_Argonaut_Encode.encodeJson(encodeJson))(bpa));
};
var decodeJsonBlogPost = new Data_Argonaut_Decode.DecodeJson(function (json) {
    return Prelude.bind(Data_Either.bindEither)(Data_Argonaut_Decode.decodeJson(Data_Argonaut_Decode.decodeStrMap(Data_Argonaut_Decode.decodeJsonJson))(json))(function (v) {
        return Prelude.bind(Data_Either.bindEither)(Data_Argonaut_Combinators[".?"](Data_Argonaut_Decode.decodeJsonInt)(v)("id"))(function (v1) {
            return Prelude.bind(Data_Either.bindEither)(Data_Argonaut_Combinators[".?"](Data_Argonaut_Decode.decodeJsonString)(v)("title"))(function (v2) {
                return Prelude.bind(Data_Either.bindEither)(Data_Argonaut_Combinators[".?"](Data_Argonaut_Decode.decodeJsonString)(v)("categories"))(function (v3) {
                    return Prelude.bind(Data_Either.bindEither)(Data_Argonaut_Combinators[".?"](Data_Argonaut_Decode.decodeJsonString)(v)("content"))(function (v4) {
                        return Prelude.pure(Data_Either.applicativeEither)({
                            id: v1, 
                            title: v2, 
                            categories: v3, 
                            content: v4
                        });
                    });
                });
            });
        });
    });
});
var decodeBlogPostArray = function (json) {
    return Prelude[">>="](Data_Either.bindEither)(Data_Argonaut_Decode.decodeJson(Data_Argonaut_Decode.decodeArray(Data_Argonaut_Decode.decodeJsonJson))(json))(Data_Traversable.traverse(Data_Traversable.traversableArray)(Data_Either.applicativeEither)(Data_Argonaut_Decode.decodeJson(decodeJsonBlogPost)));
};
module.exports = {
    BlogPost: BlogPost, 
    encodeBlogPostArray: encodeBlogPostArray, 
    decodeBlogPostArray: decodeBlogPostArray, 
    initialModel: initialModel, 
    decodeJsonBlogPost: decodeJsonBlogPost, 
    encodeJson: encodeJson
};
