// Data.Validation/index.js

/**
 *  | This module defines an applicative functor for _applicative validation_.
 *  |
 *  | Applicative validation differs from monadic validation using `Either` in
 *  | that it allows us to collect multiple errors using a `Semigroup`, whereas
 *  | `Either` terminates on the first error.
 */
"use strict";
var Prelude = require("Prelude");

/**
 *  | The `V` functor, used for applicative validation
 *  |
 *  | The `Applicative` instance collects multiple failures in
 *  | an arbitrary `Semigroup`.
 *  |
 *  | For example:
 *  |
 *  | ```purescript
 *  | validate :: Person -> V (Array Error) Person
 *  | validate person = { first: _, last: _, email: _ }
 *  |   <$> validateName person.first
 *  |   <*> validateName person.last
 *  |   <*> validateEmail person.email
 *  | ```
 */
var Invalid = (function () {
    function Invalid(value0) {
        this.value0 = value0;
    };
    Invalid.create = function (value0) {
        return new Invalid(value0);
    };
    return Invalid;
})();

/**
 *  | The `V` functor, used for applicative validation
 *  |
 *  | The `Applicative` instance collects multiple failures in
 *  | an arbitrary `Semigroup`.
 *  |
 *  | For example:
 *  |
 *  | ```purescript
 *  | validate :: Person -> V (Array Error) Person
 *  | validate person = { first: _, last: _, email: _ }
 *  |   <$> validateName person.first
 *  |   <*> validateName person.last
 *  |   <*> validateEmail person.email
 *  | ```
 */
var Valid = (function () {
    function Valid(value0) {
        this.value0 = value0;
    };
    Valid.create = function (value0) {
        return new Valid(value0);
    };
    return Valid;
})();
var showV = function (dictShow) {
    return function (dictShow1) {
        return new Prelude.Show(function (v) {
            if (v instanceof Invalid) {
                return "Invalid (" + (Prelude.show(dictShow)(v.value0) + ")");
            };
            if (v instanceof Valid) {
                return "Valid (" + (Prelude.show(dictShow1)(v.value0) + ")");
            };
            throw new Error("Failed pattern match at Data.Validation line 48, column 3 - line 49, column 3: " + [ v.constructor.name ]);
        });
    };
};

/**
 *  | Unpack the `V` type constructor, providing functions to handle the error
 *  | and success cases.
 */
var runV = function (v) {
    return function (v1) {
        return function (v2) {
            if (v2 instanceof Invalid) {
                return v(v2.value0);
            };
            if (v2 instanceof Valid) {
                return v1(v2.value0);
            };
            throw new Error("Failed pattern match at Data.Validation line 39, column 1 - line 40, column 1: " + [ v.constructor.name, v1.constructor.name, v2.constructor.name ]);
        };
    };
};

/**
 *  | Test whether validation was successful or not
 */
var isValid = function (v) {
    if (v instanceof Valid) {
        return true;
    };
    return false;
};

/**
 *  | Fail with a validation error
 */
var invalid = Invalid.create;
var functorV = new Prelude.Functor(function (v) {
    return function (v1) {
        if (v1 instanceof Invalid) {
            return new Invalid(v1.value0);
        };
        if (v1 instanceof Valid) {
            return new Valid(v(v1.value0));
        };
        throw new Error("Failed pattern match at Data.Validation line 52, column 3 - line 53, column 3: " + [ v.constructor.name, v1.constructor.name ]);
    };
});
var applyV = function (dictSemigroup) {
    return new Prelude.Apply(function () {
        return functorV;
    }, function (v) {
        return function (v1) {
            if (v instanceof Invalid && v1 instanceof Invalid) {
                return new Invalid(Prelude["<>"](dictSemigroup)(v.value0)(v1.value0));
            };
            if (v instanceof Invalid) {
                return new Invalid(v.value0);
            };
            if (v1 instanceof Invalid) {
                return new Invalid(v1.value0);
            };
            if (v instanceof Valid && v1 instanceof Valid) {
                return new Valid(v.value0(v1.value0));
            };
            throw new Error("Failed pattern match at Data.Validation line 56, column 3 - line 57, column 3: " + [ v.constructor.name, v1.constructor.name ]);
        };
    });
};
var applicativeV = function (dictSemigroup) {
    return new Prelude.Applicative(function () {
        return applyV(dictSemigroup);
    }, Valid.create);
};
module.exports = {
    isValid: isValid, 
    runV: runV, 
    invalid: invalid, 
    showV: showV, 
    functorV: functorV, 
    applyV: applyV, 
    applicativeV: applicativeV
};
